src/slots.rs

```diff
@@ -0,0 +1,110 @@
+//! Statistics of skipped and validated slots.
+
+use log::{debug, log_enabled, Level};
+use prometheus_exporter::prometheus::IntCounterVec;
+use solana_client::{client_error::ClientError, rpc_client::RpcClient};
+use solana_sdk::epoch_info::EpochInfo;
+use std::collections::BTreeMap;
+
+/// The monitor of skipped and validated slots per validator with minimal internal state.
+pub struct SkippedSlotsMonitor<'a> {
+    /// Shared Solana RPC client.
+    client: &'a RpcClient,
+    /// Prometheus gauge.
+    leader_slots: &'a IntCounterVec,
+    /// The last observed epoch number.
+    epoch_number: u64,
+    /// The last observed slot index.
+    slot_index: u64,
+    /// The slot leader schedule for the last observed epoch.
+    slot_leaders: BTreeMap<usize, String>,
+    /// `true` iff `SkippedSlotMonitor::export_skipped_slots` already ran.
+    already_ran: bool,
+}
+
+impl<'a> SkippedSlotsMonitor<'a> {
+    /// Constructs a monitor given `client`.
+    pub fn new(client: &'a RpcClient, leader_slots: &'a IntCounterVec) -> Self {
+        Self {
+            client,
+            leader_slots,
+            epoch_number: 0,
+            slot_index: 0,
+            slot_leaders: Default::default(),
+            already_ran: false,
+        }
+    }
+
+    /// Exports the skipped slot statistics given `epoch_info`.
+    pub fn export_skipped_slots(&mut self, epoch_info: &EpochInfo) -> Result<(), ClientError> {
+        let first_slot = epoch_info.absolute_slot - epoch_info.slot_index;
+
+        if self.epoch_number != epoch_info.epoch {
```

Comment:
The if-conditional here is the "happy path" (function continues to run). Could this block be inverted so that only when `self.slot_index == epoch_info.slot_index` the conditional block is entered?

Reply:
We have another case when the if-then-else-if is skipped and the function runs:
```rust
self.epoch_number == epoch_info.epoch && self.slot_index != epoch_info.slot_index
```
But we can definitely change the order of if and else here to return earlier.

Reply:
After some thought, I think the current branch order is better because updating the epoch number should have higher priority. `self.slot_index` can be from the previous epoch while `epoch_info.slot_index` can be from the current epoch.

Reply:
Ah, I see. Yes, in general if we can keep an exceptional case inside a block then that would be the best.

Reply:
OK, leaving if-else-if as is in that case. I moved `let first_slot` after that block because we don't need it if we return early.